In [2]:
import numpy as np
import pandas as pd

In [8]:
tt = pd.read_csv("https://raw.githubusercontent.com/carlson9/KocPython2020/master/homework/immSurvey.csv")
tt.tail()

Unnamed: 0   ids  X.1                                         textToSend  \
335         337  6939  337  those people that are here illegally should be...   
336         338  6940  338  racism & xenophobia make me worried!  also, co...   
337         339  6941  339  that immigrants are the heart and soul and the...   
338         340  6942  340                      job security for our citizens   
339         341  6943  341  public safety, effect on the economy, whether ...   

        Means  stanMeans    X  MetaID  treatment  pid_rep  ...    RunAll  \
335  0.650000  -0.326699  336       0          0      0.5  ... -0.131422   
336  0.650000   0.789189  337       0          1      0.5  ...  0.216873   
337  0.200000  -2.239257  338       0          0      0.5  ... -0.794429   
338  0.300000  -1.716588  339       0          1      0.5  ... -0.063889   
339  0.842105   4.268994  340       0          1      0.5  ...  1.100788   

     mergeIDs  stanMeansNewSys  stanSDsNewSys  coderMeans  newDocIDS  \
335       302        -0.507468       0.283887         0.0      10666   
336       243         0.492357       0.275906         1.5      10667   
337       264         0.246976       0.282674         0.5      10668   
338       164        -0.076251       0.315746         1.0      10669   
339       241         1.109999       0.346225         1.5      10670   

     stanMeansNewSys2  stanSDsNewSys2  stanMeansNewSysPooled  \
335         -0.558509        0.309765              -0.550042   
336          0.785079        0.373040               0.675746   
337         -0.032441        0.320736               0.043779   
338         -0.493447        0.322198              -0.234177   
339          0.185314        0.298333               0.730307   

     stanSDsNewSysPooled  
335             0.224904  
336             0.228238  
337             0.221441  
338             0.228400  
339             0.230005  

[5 rows x 38 columns]

In [9]:
alphas = tt.stanMeansNewSysPooled
sample = tt.textToSend

from sklearn.feature_extraction.text import CountVectorizer
vec = CountVectorizer()
X = vec.fit_transform(sample)
X

<340x1475 sparse matrix of type '<class 'numpy.int64'>'
	with 6239 stored elements in Compressed Sparse Row format>

In [10]:
pd.DataFrame(X.toarray(), columns=vec.get_feature_names())

11  12  125  18  1b  600  95  able  abolition  aboration  ...  wreckless  \
0     0   0    0   0   0    0   0     0          0          0  ...          0   
1     0   0    0   0   0    0   0     0          0          0  ...          0   
2     0   0    0   0   0    0   0     0          0          0  ...          0   
3     0   0    0   0   0    0   0     0          0          0  ...          0   
4     0   0    0   0   0    0   0     0          0          0  ...          0   
..   ..  ..  ...  ..  ..  ...  ..   ...        ...        ...  ...        ...   
335   0   0    0   0   0    0   0     0          0          0  ...          0   
336   0   0    0   0   0    0   0     0          0          0  ...          0   
337   0   0    0   0   0    0   0     0          0          0  ...          0   
338   0   0    0   0   0    0   0     0          0          0  ...          0   
339   0   0    0   0   0    0   0     0          0          0  ...          0   

     wrong  wrongly  xenophobia  years  yet  you  young  your  yrars  
0        0        0           0      0    0    0      0     0      0  
1        0        0           0      0    0    1      0     0      0  
2        0        0           0      0    0    0      0     0      0  
3        0        0           0      0    0    0      0     0      0  
4        0        0           0      0    0    0      0     0      0  
..     ...      ...         ...    ...  ...  ...    ...   ...    ...  
335      0        0           0      0    0    0      0     0      0  
336      0        0           1      0    0    0      0     0      0  
337      1        0           0      2    0    0      0     0      0  
338      0        0           0      0    0    0      0     0      0  
339      0        0           0      0    0    0      0     0      0  

[340 rows x 1475 columns]

In [11]:
#down-weighting frequent words; term frequency–inverse document frequency (TF–IDF), which weights the word counts by a measure of how often they appear in the documents
from sklearn.feature_extraction.text import TfidfVectorizer
vec = TfidfVectorizer()
X = vec.fit_transform(sample)
pd.DataFrame(X.toarray(), columns=vec.get_feature_names())

11   12  125   18   1b  600   95  able  abolition  aboration  ...  \
0    0.0  0.0  0.0  0.0  0.0  0.0  0.0   0.0        0.0        0.0  ...   
1    0.0  0.0  0.0  0.0  0.0  0.0  0.0   0.0        0.0        0.0  ...   
2    0.0  0.0  0.0  0.0  0.0  0.0  0.0   0.0        0.0        0.0  ...   
3    0.0  0.0  0.0  0.0  0.0  0.0  0.0   0.0        0.0        0.0  ...   
4    0.0  0.0  0.0  0.0  0.0  0.0  0.0   0.0        0.0        0.0  ...   
..   ...  ...  ...  ...  ...  ...  ...   ...        ...        ...  ...   
335  0.0  0.0  0.0  0.0  0.0  0.0  0.0   0.0        0.0        0.0  ...   
336  0.0  0.0  0.0  0.0  0.0  0.0  0.0   0.0        0.0        0.0  ...   
337  0.0  0.0  0.0  0.0  0.0  0.0  0.0   0.0        0.0        0.0  ...   
338  0.0  0.0  0.0  0.0  0.0  0.0  0.0   0.0        0.0        0.0  ...   
339  0.0  0.0  0.0  0.0  0.0  0.0  0.0   0.0        0.0        0.0  ...   

     wreckless     wrong  wrongly  xenophobia     years  yet       you  young  \
0          0.0  0.000000      0.0    0.000000  0.000000  0.0  0.000000    0.0   
1          0.0  0.000000      0.0    0.000000  0.000000  0.0  0.253915    0.0   
2          0.0  0.000000      0.0    0.000000  0.000000  0.0  0.000000    0.0   
3          0.0  0.000000      0.0    0.000000  0.000000  0.0  0.000000    0.0   
4          0.0  0.000000      0.0    0.000000  0.000000  0.0  0.000000    0.0   
..         ...       ...      ...         ...       ...  ...       ...    ...   
335        0.0  0.000000      0.0    0.000000  0.000000  0.0  0.000000    0.0   
336        0.0  0.000000      0.0    0.195271  0.000000  0.0  0.000000    0.0   
337        0.0  0.078393      0.0    0.000000  0.150361  0.0  0.000000    0.0   
338        0.0  0.000000      0.0    0.000000  0.000000  0.0  0.000000    0.0   
339        0.0  0.000000      0.0    0.000000  0.000000  0.0  0.000000    0.0   

     your  yrars  
0     0.0    0.0  
1     0.0    0.0  
2     0.0    0.0  
3     0.0    0.0  
4     0.0    0.0  
..    ...    ...  
335   0.0    0.0  
336   0.0    0.0  
337   0.0    0.0  
338   0.0    0.0  
339   0.0    0.0  

[340 rows x 1475 columns]

In [14]:
#https://www.google.com/search?q=train_test_split+scikit&oq=train_test_split+scikit
from sklearn.model_selection import train_test_split
Xtrain, Xtest, ytrain, ytest = train_test_split(X, alphas,
random_state=1)
from sklearn.gaussian_process import GaussianProcessRegressor
from sklearn.gaussian_process.kernels import ConstantKernel, RBF

rbf = ConstantKernel(1.0) * RBF(length_scale=1.0)
gpr = GaussianProcessRegressor(kernel=rbf, alpha=1e-8)

gpr.fit(Xtrain.toarray(), ytrain)

mu_s, cov_s = gpr.predict(Xtest.toarray(), return_cov=True)

np.corrcoef(ytest, mu_s)

array([[1.       , 0.6832852],
       [0.6832852, 1.       ]])

### Correlation Coefficient is 0.6832852.

In [18]:
bigram_vectorizer =  CountVectorizer(ngram_range=(1, 2), token_pattern=r'\b\w+\b', min_df=1)
analyze = bigram_vectorizer.build_analyzer()
analyze("My name is Fuat")

['my', 'name', 'is', 'fuat', 'my name', 'name is', 'is fuat']

In [19]:
bigram_vectorizer =  CountVectorizer(ngram_range=(2, 2), token_pattern=r'\b\w+\b', min_df=1)
analyze = bigram_vectorizer.build_analyzer()
analyze("My name is Fuat")

['my name', 'name is', 'is fuat']

In [16]:
X = bigram_vectorizer.fit_transform(sample)

pd.DataFrame(X.toarray(), columns=bigram_vectorizer.get_feature_names())

Xtrain, Xtest, ytrain, ytest = train_test_split(X, alphas,
random_state=1)

rbf = ConstantKernel(1.0) * RBF(length_scale=1.0)
gpr = GaussianProcessRegressor(kernel=rbf, alpha=1e-8)

gpr.fit(Xtrain.toarray(), ytrain)

mu_s, cov_s = gpr.predict(Xtest.toarray(), return_cov=True)

np.corrcoef(ytest, mu_s)

array([[1.        , 0.45545545],
       [0.45545545, 1.        ]])

### Using the bigram vectorizer reduced the corr. coef. r from 0.6832852 to 0.45545545